craft_parts/executor/environment.py

```diff
@@ -244,6 +244,12 @@ def expand_environment(
         replacements[f"${key}"] = value
         replacements[f"${{{key}}}"] = value
 
+    # order is important - for example, `CRAFT_ARCH_TRIPLET_BUILD_{ON|FOR}` should be
+    # evaluated before `CRAFT_ARCH_TRIPLET` to avoid premature variable expansion
+    replacements = dict(
```

Comment:
I've been trying to come up with a falsifying example here, since it makes me mildly uncomfortable and I'm not sure why. This might be out of scope, but here's the best I've got:

```
...
override-build: |
  export CRAFT_ARCH_TRIPLET_VALID=$(some_command_output)
  if [ "$CRAFT_ARCH_TRIPLET_VALID" == "false" ]; then
    exit 1
  fi
```

This was still an issue before, related to the fact that we were doing a simple string replacement rather than finding either `$KEY` at word boundaries or a full `${KEY}` whether it's at a word boundary or not.

This PR fixes the bug it's intended for so I'm approving, but we might want to discuss whether we care about the example I provided, especially since there's a clear workaround (use the curly braces, Luke!)

Reply:
Yeah, you're right.  This corner case still is possible.

@tigarmo - is this why you were surprised that project variables were evaluated inside `override` scripts?

A possible solution for your example is to not expand project variables inside `override` scripts.  Instead, it would be done at runtime (when the override script is executed and the project variables are in the build environment).  Bash would evaluate `$CRAFT_ARCH_TRIPLET` and `$CRAFT_ARCH_TRIPLET_VALID` at the same time and resolve this corner case.